### PR TITLE
Ingest/output STAC items with EO attributes

### PR DESCRIPTION
**Pull request by mcovalt**
Closes https://github.com/ResonantGeoData/RGD-ScrumBoard/issues/16
<details>
<summary>Example JSON</summary>
<p>

```json
{
  "type": "Feature",
  "stac_version": "1.0.0-beta.2",
  "id": 22,
  "properties": {
    "datetime": "2021-07-13T16:29:20.464696Z",
    "platform": null,
    "proj:epsg": 32613,
    "proj:transform": [
      313785.0,
      1000.0,
      0.0,
      4583115.0,
      0.0,
      -1000.0
    ],
    "eo:bands": []
  },
  "geometry": {
    "type": "Polygon",
    "coordinates": [
      [
        [
          -107.2268691156271,
          41.37802031673873
        ],
        [
          -104.44032726896141,
          41.39821189086375
        ],
        [
          -104.45769347758004,
          39.26311024561664
        ],
        [
          -107.15783327662788,
          39.24437827343882
        ],
        [
          -107.2268691156271,
          41.37802031673873
        ]
      ]
    ]
  },
  "links": [],
  "assets": {
    "image-40": {
      "href": "https://data.kitware.com/api/v1/file/hashsum/sha512/b0b52a537d79460afa63a4849c2c03cf686b6f32446a2c56320027f7e701965b0f2af31e0dd843471bc98eecd5f0dd1be67d8c38b4759a81a5a0aa707ae4fea6/download",
      "title": "LC08_L1TP_034032_20200429_20200509_01_T1_sr_band1.tif",
      "eo:bands": [
        {
          "name": "B1",
          "common_name": "nir"
        }
      ],
      "roles": [
        "data"
      ]
    },
    "image-41": {
      "href": "https://data.kitware.com/api/v1/file/hashsum/sha512/492e30b6fdeebf67332d87ab07258c6bed1f9830c214492866bbd2f06f53fcba72b4984709cd4a486f49a4f1a5effaf482cda95f1456d6e3b7e693bee8d9c200/download",
      "title": "LC08_L1TP_034032_20200429_20200509_01_T1_sr_band2.tif",
      "eo:bands": [
        {
          "name": "B2",
          "common_name": "rededge"
        }
      ],
      "roles": [
        "data"
      ]
    },
    "image-42": {
      "href": "https://data.kitware.com/api/v1/file/hashsum/sha512/adf8114c240ad6d5462fcac0b480cabc827da3b095620741c4eb414da400c368a8f112a6f13470870e86cbefed6235efac6c3611bd3efd53dd6e12114006af56/download",
      "title": "LC08_L1TP_034032_20200429_20200509_01_T1_sr_band3.tif",
      "eo:bands": [
        {
          "name": "B3",
          "center_wavelength": 0.42,
          "full_width_half_max": 0.02
        }
      ],
      "roles": [
        "data"
      ]
    },
    "ancillary-49": {
      "href": "https://data.kitware.com/api/v1/file/hashsum/sha512/e19e0c0c28c67441700cf272cb6ae20e5cc0baee24e5527e096e61e290ca823913224cdbabb884c5550e73587192428c0650921a00630c82f45c4eddf52c652f/download",
      "title": "stars.png",
      "roles": [
        "metadata"
      ]
    }
  },
  "bbox": [
    313785.0,
    4346115.0,
    546785.0,
    4583115.0
  ],
  "stac_extensions": [
    "projection",
    "eo"
  ]
}
```

</p>
</details>

### CODE REVIEW

**Review by banesullivan — APPROVED** [review]
Overall looks great! Only a few comments

File: django-rgd-imagery/rgd_imagery/models/base.py [hunk 1]
```diff
@@ -36,7 +37,9 @@ class ImageMeta(TimeStampedModel, PermissionPathMixin):
     driver = models.CharField(max_length=100)
     height = models.PositiveIntegerField()
     width = models.PositiveIntegerField()
-    number_of_bands = models.PositiveIntegerField()
+    number_of_bands = (
+        models.PositiveIntegerField()
+    )  # TODO: code smell? this can be computed relationally
```
[the remark this answers is not in the document]
Comment (mcovalt) [reply django-rgd-imagery/rgd_imagery/models/base.py:42]: Sounds good. I've added a new issue to address this: https://github.com/ResonantGeoData/ResonantGeoData/issues/478.

File: django-rgd-imagery/rgd_imagery/serializers/stac.py [hunk 2]
```diff
@@ -91,24 +132,46 @@ def to_representation(self, instance: models.RasterMeta) -> dict:
...
+                    elif eo_band.center_wavelength and eo_band.full_width_half_max:
+                        eo_band_spectral_lower = (
+                            eo_band.eo_band_spectral_upper - eo_band.full_width_half_max / 2
+                        )
+                        eo_band_spectral_upper = (
+                            eo_band.center_wavelength + eo_band.full_width_half_max / 2
+                        )
+                    models.BandMeta.objects.get_or_create(
+                        parent_image=image,
+                        band_number=eo_band_number,
```
[the remark this answers is not in the document]
Comment (mcovalt) [reply django-rgd-imagery/rgd_imagery/serializers/stac.py:162]: Good catch! Thanks. What you describe is the intended behavior. I've implemented it in https://github.com/ResonantGeoData/ResonantGeoData/pull/471/commits/cc55d7da358a1b38ed154d4c06ce5819fe5ab1e9.

File: django-rgd-imagery/tests/data_fixtures.py [hunk 3]
```diff
@@ -132,14 +134,24 @@ def sample_raster_url():
...
             file__url=datastore.get_url(f),
         )
-        for f in landsat_files
-    ]
+        images.append(image)
+        factories.BandMetaFactory(
+            parent_image=image,
+            band_range=band_range,
+            band_number=(i + 1),
+        )
```
[the remark this answers is not in the document]
Comment (mcovalt) [reply django-rgd-imagery/tests/data_fixtures.py:154]: > This test isn't a realistic use case as there are two `(i+1)` `BandMetas` here.

To be sure we're on the same page: this creates three `BandMeta` objects with `band_number`s of 1, 2, and 3. Each `BandMeta` creates special edge-cases to test (mainly in `band_ranges`).

> Anywhere we use the `Meta` suffix, we ought not manually create these records. `*Meta` models should all be managed by the ETL routines.

This makes sense, but it seems like it would be difficult to manufacture all the edge-cases that directly creating `BandMeta` has allowed. Is there an easy way to create an `Image` via `ImageFactory` with the band information here?

[the remark this answers is not in the document]
Comment (banesullivan) [reply django-rgd-imagery/tests/data_fixtures.py:154]: > but it seems like it would be difficult to manufacture all the edge-cases that directly creating BandMeta has allowed

definitely difficult... 

> Is there an easy way to create an Image via ImageFactory with the band information here?

No, not really... because the ETL routines populate these. Perhaps the best thing to do here is just to update the existing `BandMeta` records rather than creating new ones.

[the remark this answers is not in the document]
Comment (banesullivan) [reply django-rgd-imagery/tests/data_fixtures.py]: I've gone ahead and added this change. There's really no good way to do this at the moment because the ETL routines do not populate all of the fields on the `BandMeta` model. We can come back to this if we need to.

**Comment by banesullivan** [comment]
Awesome work, @mcovalt! Merging...
